### ISSUE TITLE
Segfaults in v2.0.{0,1}

### ISSUE THREAD

**Issue by janosh** [spglib/spglib#197]
## Describe the bug

A number of `pymatgen` operations started crashing with segfaults after updating to v2.0.0.

## To Reproduce

```sh
pip install spglib==2.0.0  # or 2.0.1
```

Run below code with `python -Xfaulthandler repro.py`:

```py
# repro.py
import faulthandler

import spglib
from pymatgen.entries.compatibility import MaterialsProjectCompatibility
from pymatgen.entries.computed_entries import ComputedStructureEntry

print(f"{spglib.__version__=}")

faulthandler.enable()

dct = {
    "@module": "pymatgen.entries.computed_entries",
    "@class": "ComputedStructureEntry",
    "energy": -32.50022431,
    "composition": {"S": 2.0, "Bi": 1.0, "Li": 6.0},
    "correction": 0.0,
    "parameters": {
        "potcar_symbols": [
            "PAW_PBE S 17Jan2003",
            "PAW_PBE Bi 08Apr2002",
            "PAW_PBE Li_sv 23Jan2001",
        ],
        "is_hubbard": False,
        "hubbards": {},
        "run_type": "GGA",
        "potcar_spec": [
            {"hash": None, "titel": "PAW_PBE S 17Jan2003"},
            {"hash": None, "titel": "PAW_PBE Bi 08Apr2002"},
            {"hash": None, "titel": "PAW_PBE Li_sv 23Jan2001"},
        ],
    },
    "data": {},
    "entry_id": "wbm-step-1-24459",
    "structure": {
        "@module": "pymatgen.core.structure",
        "@class": "Structure",
        "charge": None,
        "lattice": {
            "matrix": [
                [4.0332905, 0.00019654, 0.00033189],
                [0.00019791, 4.09449716, 0.00105079],
                [2.01755058, 2.04667974, 9.83595687],
            ],
            "a": 4.0332905184,
            "b": 4.0944972996,
            "c": 10.2472169807,
            "alpha": 78.4638510747,
            "beta": 78.6397987424,
            "gamma": 89.9944373726,
            "volume": 162.4224924451,
        },
        "sites": [
            {
                "species": [{"element": "S", "occu": 1}],
                "abc": [0.67847964, 0.67665944, 0.64561893],
                "xyz": [4.0392082508, 4.0920886875, 6.3512161575],
                "label": "S",
                "properties": {},
            },
            {
                "species": [{"element": "S", "occu": 1}],
                "abc": [0.32476989, 0.32232743, 0.35455337000000003],
                "xyz": [2.0252844611, 2.0454897761, 3.4878181418],
                "label": "S",
                "properties": {},
            },
            {
                "species": [{"element": "Bi", "occu": 1}],
                "abc": [0.00178017, 0.99953694, 8.554e-05],
                "xyz": [0.0075503424, 4.092776585, 0.001892262],
                "label": "Bi",
                "properties": {},
            },
            {
                "species": [{"element": "Li", "occu": 1}],
                "abc": [0.60702988, 0.10487906, 0.78886588],
                "xyz": [4.0399254186, 2.0441021331, 7.7595624449],
                "label": "Li",
                "properties": {},
            },
            {
                "species": [{"element": "Li", "occu": 1}],
                "abc": [0.10382818, 0.60266268, 0.79410993],
                "xyz": [2.0210454349, 4.0929097432, 7.811498753],
                "label": "Li",
                "properties": {},
            },
            {
                "species": [{"element": "Li", "occu": 1}],
                "abc": [0.39640045, 0.89413866, 0.21137099],
                "xyz": [2.0254267916, 4.0937348354, 2.0801070545],
                "label": "Li",
                "properties": {},
            },
            {
                "species": [{"element": "Li", "occu": 1}],
                "abc": [0.7519066999999999, 0.24963494, 0.49971919000000004],
                "xyz": [4.0409162969, 2.0450424745, 4.9157282642],
                "label": "Li",
                "properties": {},
            },
            {
                "species": [{"element": "Li", "occu": 1}],
                "abc": [0.25116892, 0.74921193, 0.5004725999999999],
                "xyz": [2.0229142799, 4.0920026152, 4.9234975331],
                "label": "Li",
                "properties": {},
            },
            {
                "species": [{"element": "Li", "occu": 1}],
                "abc": [0.89971318, 0.39648191, 0.20599457000000002],
                "xyz": [4.0444875535, 2.0451757971, 2.026868931],
                "label": "Li",
                "properties": {},
            },
        ],
    },
}

cse = ComputedStructureEntry.from_dict(dct)

MaterialsProjectCompatibility().process_entry(cse)
```
</details>

## Expected behavior

No crash.

## Environment

- Python version: 3.10.8
- Spglib's version: 2.0.1

## Additional context

Related: #194, https://github.com/microsoft/vscode-jupyter/issues/11808

New issues prompted by https://github.com/spglib/spglib/issues/194#issuecomment-1297802841.

**Comment by atztogo** [comment]
I don't use pymatgen and don't know how it uses spglib. Couid it be possible for you to write a python script using python-spgilb?
https://spglib.github.io/spglib/python-spglib.html#crystal-structure-cell
https://spglib.github.io/spglib/python-spglib.html#get-symmetry-dataset

**Comment by janosh** [comment]
Did I install the `develop` branch correctly in https://github.com/spglib/spglib/issues/194#issuecomment-1297734990? I might wait for the next release, first try and that and if the issue really persists, I'll create another repro.

**Comment by lan496** [comment]
@janosh Can you check a commit of your installed `develop` branch?
Can you also try `spglib.get_symmetry_dataset` directly, and report the structure in [spglib's format](https://spglib.github.io/spglib/python-spglib.html#crystal-structure-cell)?
A snippet is shown at https://github.com/spglib/spglib/issues/193#issuecomment-1275393012.

**Comment by janosh** [comment]
@lan496 I installed the dev branch at 36d570b (Merge pull request #196 from lan496/fix-segfault-magdataset).

```py
from pymatgen.symmetry.analyzer import SpacegroupAnalyzer

# SpacegroupAnalyzer uses symprec=0.01, angle_tolerance=5.0 as default
spa = SpacegroupAnalyzer(cse.structure)
cell = spa._cell
print(cell)
dataset = spglib.get_symmetry_dataset(cell)
print(dataset)
```


`structure`:

```
(array([[4.03329050e+00, 1.96540000e-04, 3.31890000e-04],
       [1.97910000e-04, 4.09449716e+00, 1.05079000e-03],
       [2.01755058e+00, 2.04667974e+00, 9.83595687e+00]]), array([[6.7847964e-01, 6.7665944e-01, 6.4561893e-01],
       [3.2476989e-01, 3.2232743e-01, 3.5455337e-01],
       [1.7801700e-03, 9.9953694e-01, 8.5540000e-05],
       [6.0702988e-01, 1.0487906e-01, 7.8886588e-01],
       [1.0382818e-01, 6.0266268e-01, 7.9410993e-01],
       [3.9640045e-01, 8.9413866e-01, 2.1137099e-01],
       [7.5190670e-01, 2.4963494e-01, 4.9971919e-01],
       [2.5116892e-01, 7.4921193e-01, 5.0047260e-01],
       [8.9971318e-01, 3.9648191e-01, 2.0599457e-01]]), [1, 1, 2, 3, 3, 3, 3, 3, 3], [0, 0, 0, 0, 0, 0, 0, 0, 0])
```

`data`
```
{'number': 1, 'hall_number': 1, 'international': 'P1', 'hall': 'P 1', 'choice': '', 'transformation_matrix': array([[-1.00000000e+00, -4.92837650e-17, -1.00000000e+00],
       [ 3.26853074e-17, -1.00000000e+00, -1.00000000e+00],
       [ 0.00000000e+00,  1.35525272e-20,  1.00000000e+00]]), 'origin_shift': array([0., 0., 0.]), 'rotations': array([[[1, 0, 0],
        [0, 1, 0],
        [0, 0, 1]]], dtype=int32), 'translations': array([[0., 0., 0.]]), 'wyckoffs': ['a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a'], 'site_symmetry_symbols': ['1', '1', '1', '1', '1', '1', '1', '1', '1'], 'crystallographic_orbits': array([0, 1, 2, 3, 4, 5, 6, 7, 8], dtype=int32), 'equivalent_atoms': array([0, 1, 2, 3, 4, 5, 6, 7, 8], dtype=int32), 'primitive_lattice': array([[-4.03329050e+00, -1.96540000e-04, -3.31890000e-04],
       [ 1.97910000e-04,  4.09449716e+00,  1.05079000e-03],
       [ 2.01593783e+00,  2.04801396e+00, -9.83457419e+00]]), 'mapping_to_primitive': array([0, 1, 2, 3, 4, 5, 6, 7, 8], dtype=int32), 'std_lattice': array([[4.03329052e+00, 0.00000000e+00, 0.00000000e+00],
       [3.97519032e-04, 4.09449728e+00, 0.00000000e+00],
       [2.01522836e+00, 2.04539180e+00, 9.83526528e+00]]), 'std_types': array([1, 1, 2, 3, 3, 3, 3, 3, 3], dtype=int32), 'std_positions': array([[6.7590143e-01, 6.7772163e-01, 6.4561893e-01],
       [3.2067674e-01, 3.2311920e-01, 3.5455337e-01],
       [9.9813429e-01, 3.7752000e-04, 8.5540000e-05],
       [6.0410424e-01, 1.0625506e-01, 7.8886588e-01],
       [1.0206189e-01, 6.0322739e-01, 7.9410993e-01],
       [3.9222856e-01, 8.9449035e-01, 2.1137099e-01],
       [7.4837411e-01, 2.5064587e-01, 4.9971919e-01],
       [2.4835848e-01, 7.5031547e-01, 5.0047260e-01],
       [8.9429225e-01, 3.9752352e-01, 2.0599457e-01]]), 'std_rotation_matrix': array([[-9.99999995e-01, -4.87294429e-05, -8.22876504e-05],
       [ 4.87505587e-05, -9.99999966e-01, -2.56626691e-04],
       [-8.22751423e-05, -2.56630702e-04,  9.99999964e-01]]), 'std_mapping_to_primitive': array([0, 1, 2, 3, 4, 5, 6, 7, 8], dtype=int32), 'pointgroup': '1'}
```

This was run with `spglib.__version__='1.16.4'` since v2.0.0 crashes.

**Comment by atztogo** [comment]
@lan496, I tested the structure given by @janosh above on M1 Mac and Ubuntu 20.04 (amd64), and both worked (or say didn't crash) with latest develop.

**Comment by janosh** [comment]
@atztogo Just to check, does it crash for you under v2?

**Comment by atztogo** [comment]
It didn't crash at the commit 36d570bc3f on my M1 mac and Ubutnu 20.04 amd64. But this doesn't mean it works for every system, and it can crash on your system.

**Comment by lan496** [comment]
I cannot reproduce the crash at the commit https://github.com/spglib/spglib/commit/36d570bc3fbeda20903bfbbb7db05e40d284a319 on my M1 mac and Ubuntu 20.04 (x86-64) too. 
But at the current master be3eca7ad6364d9d8381ad01f89649d79b44d6b0, `spglib.get_magnetic_symmetry` causes a segmentation fault on my M1 mac.
@janosh the spglib you installed might be older or not reflected in your environment.

<details><summary>script</summary>
```python
import spglib
import numpy as np

print(f"spglib=={spglib.__version__}")

# https://github.com/spglib/spglib/issues/197

# Row-wise for python!
lattice = np.array([
    [4.03329050e+00, 1.96540000e-04, 3.31890000e-04],
    [1.97910000e-04, 4.09449716e+00, 1.05079000e-03],
    [2.01755058e+00, 2.04667974e+00, 9.83595687e+00]
])
positions = np.array([
    [6.7847964e-01, 6.7665944e-01, 6.4561893e-01],
    [3.2476989e-01, 3.2232743e-01, 3.5455337e-01],
    [1.7801700e-03, 9.9953694e-01, 8.5540000e-05],
    [6.0702988e-01, 1.0487906e-01, 7.8886588e-01],
    [1.0382818e-01, 6.0266268e-01, 7.9410993e-01],
    [3.9640045e-01, 8.9413866e-01, 2.1137099e-01],
    [7.5190670e-01, 2.4963494e-01, 4.9971919e-01],
    [2.5116892e-01, 7.4921193e-01, 5.0047260e-01],
    [8.9971318e-01, 3.9648191e-01, 2.0599457e-01]
])
numbers = [1, 1, 2, 3, 3, 3, 3, 3, 3]
magmoms =  [0, 0, 0, 0, 0, 0, 0, 0, 0]

for symprec in [1e-5, 1e-4, 1e-3, 1e-2, 1e-1]:
    for angle_tolerance in [-1, 1, 2, 4]:
        print(f"{symprec=}, {angle_tolerance=}")
        symmetry = spglib.get_symmetry(
            (lattice, positions, numbers),
            symprec=symprec,
            angle_tolerance=angle_tolerance,
        )
        dataset = spglib.get_symmetry_dataset(
            (lattice, positions, numbers, magmoms),
            symprec=symprec,
            angle_tolerance=angle_tolerance
        )
        magnetic_symmetry = spglib.get_magnetic_symmetry(
            (lattice, positions, numbers, magmoms),
            symprec=symprec,
            angle_tolerance=angle_tolerance,
            mag_symprec=symprec,
        )

        assert symmetry is not None
        assert dataset is not None
        assert magnetic_symmetry is not None
        print(f"  {dataset['hall']}")

print("[Success]")
```
</details>

**Comment by janosh** [comment]
@lan496 Thanks for confirming! Would it be possible to get a new release including the fix?

**Comment by atztogo** [comment]
@lan496, is the develop branch ready for release?

**Comment by lan496** [comment]
@atztogo Yes, I'm creating PR for the next patch version.

**Comment by atztogo** [comment]
Released v2.0.2. The conda-forge package will be available later soon.

**Comment by janosh** [comment]
Thanks! Fix confirmed.
